python/ray/tests/test_scheduling_2.py

```diff
@@ -0,0 +1,145 @@
+import os
+import sys
+import tempfile
+
+import ray
+
+from ray._private.test_utils import (Semaphore)
+
+
```

Comment:
Can you add a simple barrier test like this?

```
import ray
import time
import asyncio

ray.init(num_cpus=2)


@ray.remote(num_cpus=0)
class Barrier:
    def __init__(self, limit):
        self.i = 0
        self.limit = limit

    async def join(self):
        self.i += 1
        while self.i < self.limit:
            await asyncio.sleep(1)


b = Barrier.remote(10)

@ray.remote
def f(i, start):
    print("Launch", i, time.time() - start)
    ray.get(b.join.remote())

start = time.time()
ray.get([f.remote(i, start) for i in range(10)])
```

Btw, when I ran the above, I expected to see exponential scale-up of the delays, but I got this instead:
```
(f pid=155291) Launch 0 0.03812384605407715
(f pid=155362) Launch 1 0.5240130424499512
(f pid=155361) Launch 2 9.340767860412598
(f pid=155424) Launch 3 9.78423523902893
(f pid=155423) Launch 4 9.785459518432617
(f pid=155471) Launch 5 19.341200351715088
(f pid=155509) Launch 6 19.79110550880432
(f pid=155508) Launch 7 19.798838138580322
```

Followed by hanging forever. I think we should have a max cap on the time delay, and also figure out why the step-up isn't smooth above (e.g., 1, 2, 4, 8, etc.)

Reply:
Added as a test and just reran with latest changes pulled in, which seem to have fixed it. 

```
(f pid=90535)[0m Launch 0 0.05490303039550781
(f pid=90543)[0m Launch 1 0.6045780181884766
(f pid=90542)[0m Launch 2 0.6046791076660156
(f pid=90548)[0m Launch 3 1.4007248878479004
(f pid=90552)[0m Launch 4 1.941673994064331
(f pid=90551)[0m Launch 5 2.2514801025390625
(f pid=90557)[0m Launch 6 3.883697032928467
(f pid=90560)[0m Launch 7 7.1504950523376465
(f pid=90563)[0m Launch 8 13.55370807647705
(f pid=90566)[0m Launch 9 26.449985027313232
12.89627718925476 6.403211832046509
```

The ratio between the last two numbers is 2.014. Adding as a test (I guess we can assert the last numbers grow exponentially since the short wait times are pretty noisy).